extensions/2.0/Khronos/KHR_interactivity/Specification.adoc

```diff
@@ -0,0 +1,2914 @@
+
+==== Custom Event Nodes
+
+===== Receive
+
+[cols="1h,1,2"]
+|===
+| Type | `event/receive` | Receive a custom event
+| Configuration
+| `int event` | The custom event index
+| Output value sockets
+| `<custom>` | Output values defined by the custom event
+|===
+
+This node is activated when a custom event specified by the `event` configuration value occurs. The types, names, and semantics of the output value sockets are defined by the custom event index.
+
+The `event` configuration value **MUST** be non-negative and less than the total number of custom event definitions, otherwise the node is invalid.
+
+This node has no internal state.
+
+===== Send
+
+[cols="1h,1,2"]
+|===
+| Type | `event/send` | Send a custom event
+| Configuration
+| `int event` | The custom event index
+| Input flow sockets
+| `in` | The entry flow into this node
+| Input value sockets
+| `<custom>` | Input values defined by the custom event
+.2+| Output flow sockets
+| `out` | The flow to be activated after sending the event
+| `err` | The flow to be activated if any of the input values is invalid
+|===
+
+This node sends a custom event specified by the `event` configuration value. The types, names, and validation rules of the input value sockets are defined by the custom event index.
+
+The `event` configuration value **MUST** be non-negative and less than the total number of custom event definitions, otherwise the node is invalid.
+
+This node has no internal state.
+
+When the `in` input flow is activated:
+
+1. Evaluate all input values.
+2. If any of them is invalid, activate the `err` output flow and skip the next step.
+3. Activate the `out` output flow.
+
+= JSON Syntax
+
+==  General
+
+A `KHR_interactivity` extension object is added to the root-level `extensions` property. It contains four arrays corresponding to four interactivity concepts: `types`, `events`, `variables`, and `nodes`. As with the core glTF spec, if a JSON array is empty, it **MUST** be omitted from the asset.
+
+```json
+{
+  "asset": {
+    "version": "2.0"
+  },
+  "extensionsUsed": ["KHR_interactivity"],
+  "extensions": {
+    "KHR_interactivity": {
+      "types": [
+        //
+      ],
+      "events": [
+        //
+      ],
+      "variables": [
+        //
+      ],
+      "nodes": [
+        //
+      ]
+    }
+  }
+}
+```
+
+== Types
+
+The `types` array defines mappings between graph-local type indices and the recognized type signatures.
+
+The following example defines type `0` as *float2*, type `1` as *int*, and type `2` as *float*:
+
+```json
+"types": [
+  {
+    "signature": "float2"
+  },
+  {
+    "signature": "int"
+  },
+  {
+    "signature": "float"
+  }
+]
+```
+
+The signature value **MUST** be one of the value types defined in this extension specification or `"custom"`. In the latter case, the type semantic **MUST** be provided by an additional extension.
+
+Non-custom signature **MUST NOT** appear more than once.
+
+== Events
```

Comment:
Can we get a high-level explanation of how this might work with a trigger volume as defined by `KHR_physics_rigid_bodies`? Am I reading correctly that the graph would have a script block (graph node) with type `event/receive` that points to an item in the document-level `"events"` array? And this event could have an ID of something like `"trigger/bodyEntered"`, which is defined by the extension to emit when a body enters the trigger volume?

What would the output value type of that custom event be? It would need to be something that allows us to reference which object did the entering. A glTF node index comes to mind, but then that wouldn't work for interacting between scenes. Probably the only option here would be a runtime-determined ID for each node in the scene (could just be the memory address), which allows for many glTF files to interact with each other. However, this would then mean that in order to use this value, the `KHR_interactivity` graph would need a script block (graph node) that can access an object using the same runtime-provided IDs, similar to how object model properties can be accessed using JSON pointers.

For example, let's say we want to build a trigger volume that teleports any object that enters it to another location. Assume that the trigger volume and the destination location are already known in the glTF file. How can the interactivity graph allow objects from other glTF scenes to be teleported, or even non-glTF objects provided by the game engine at runtime, including the player object itself? If a player walks into a glTF teleporter, I expect that the glTF should be able to detect that, and set the position of the player.

Reply:
This is not a response to the above, but I have an additional question.
The Behave Graph Authoring tool allows the creation of a "node/OnSelect" node, but there is no definition of this node in the Interactivity Extension Specification.
Is this simply a documentation mistake, or is it related to KHR_node_selectability?